crates/sui-framework/sources/governance/staking_pool.move

```diff
@@ -0,0 +1,253 @@
+        assert!(
+            delegation.validator_address == pool.validator_address &&
+            delegation.pool_starting_epoch == pool.starting_epoch,
+            WRONG_POOL
+        );
+        assert!(is_delegation_activated(delegation), DELEGATION_NOT_YET_ACTIVATED);
+
+        let pool_tokens_ref = option::borrow_mut(&mut delegation.pool_tokens);
+        let pool_token_balance = balance::value(pool_tokens_ref);
+        assert!(pool_token_balance >= withdraw_pool_token_amount, INSUFFICIENT_POOL_TOKEN_BALANCE);
+
+        // Calculate the amount of SUI tokens that should be withdrawn from the pool using the current exchange rate.
+        let sui_withdraw_amount = get_sui_amount(pool, withdraw_pool_token_amount);
+
+        let sui_principal_amount = balance::value(&delegation.principal);
+
+        // Calculate the amounts if SUI to be withdrawn from the principal component and the rewards component.
+        let sui_withdraw_from_principal = (sui_principal_amount as u128) * (withdraw_pool_token_amount as u128) / (pool_token_balance as u128);
+        let sui_withdraw_from_rewards = sui_withdraw_amount - (sui_withdraw_from_principal as u64); 
+
+        // burn the pool tokens
+        balance::decrease_supply(
+            &mut pool.delegation_token_supply, 
+            balance::split(pool_tokens_ref, withdraw_pool_token_amount)
+        );
+
+        // TODO: implement withdraw bonding period here.
+        withdraw_from_principal(delegation, (sui_withdraw_from_principal as u64), ctx);
+
+        let reward_withdraw = balance::split(&mut pool.rewards_pool, sui_withdraw_from_rewards);
+        // TODO: implement withdraw bonding period here.
+        transfer::transfer(coin::from_balance(reward_withdraw, ctx), tx_context::sender(ctx));
+
+        // decrement sui balance in the pool
+        pool.sui_balance = pool.sui_balance - sui_withdraw_amount;
+    }
+
+    /// Deactivate a staking pool by wrapping it in an `InactiveStakingPool` and sharing this newly created object. 
+    /// After this pool deactivation, the pool stops earning rewards. Only delegation withdraws can be made to the pool.
+    public(friend) fun deactivate_staking_pool(pool: StakingPool, ctx: &mut TxContext) {
+        let inactive_pool = InactiveStakingPool { id: object::new(ctx), pool};
+        // TODO: emit events here with the id of the inactive pool.
+        transfer::share_object(inactive_pool);
+    }
+
+    /// Withdraw delegation from an inactive pool.
+    public entry fun withdraw_from_inactive_pool(
+        inactive_pool: &mut InactiveStakingPool, 
+        delegation: &mut Delegation, 
+        withdraw_amount: u64, 
+        ctx: &mut TxContext
+    ) {
+        withdraw_stake(&mut inactive_pool.pool, delegation, withdraw_amount, ctx);
+    }
+
+    /// Destroy an empty delegation that no longer contains any SUI or pool tokens.
+    public entry fun destroy_empty_delegation(delegation: Delegation) {
```

Comment:
I know it's a draft, but something to keep in mind - we actually need to check every value here and emit a custom error to not misguide people when they receive an error from `balance` module or `option`.